stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/file/Delete.java

```diff
@@ -24,53 +24,56 @@
 import org.ballerinalang.natives.annotations.BallerinaFunction;
-import org.ballerinalang.natives.annotations.Receiver;
-import org.ballerinalang.util.exceptions.BallerinaException;
+import org.ballerinalang.natives.annotations.ReturnType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Comparator;
+
+import static org.ballerinalang.nativeimpl.file.utils.FileUtils.createAccessDeniedError;
+import static org.ballerinalang.nativeimpl.file.utils.FileUtils.createIOError;
 
 /**
  * Deletes a file from a given location.
  */
 @BallerinaFunction(
         packageName = "ballerina.file",
         functionName = "delete",
-        receiver = @Receiver(type = TypeKind.STRUCT, structType = "File",
-                             structPackage = "ballerina.file"),
         args = {@Argument(name = "target", type = TypeKind.STRUCT, structType = "File",
                 structPackage = "ballerina.file")},
+        returnType = {
+                @ReturnType(type = TypeKind.STRUCT, structType = "AccessDeniedError", structPackage = "ballerina.file"),
+                @ReturnType(type = TypeKind.STRUCT, structType = "IOError", structPackage = "ballerina.file")},
         isPublic = true
 )
 public class Delete extends AbstractNativeFunction {
 
+    private static final Logger log = LoggerFactory.getLogger(Delete.class);
+
     @Override 
     public BValue[] execute(Context context) {
-
         BStruct target = (BStruct) getRefArgument(context, 0);
-        File targetFile = new File(target.getStringField(0));
-        if (!targetFile.exists()) {
-            throw new BallerinaException("failed to delete file: file not found: " + targetFile.getPath());
-        }
-        if (!delete(targetFile)) {
-            throw new BallerinaException("failed to delete file: " + targetFile.getPath());
-        }
-        return VOID_RETURN;
-    }
+        Path targetFile = Paths.get(target.getStringField(0));
 
-    private boolean delete(File targetFile) {
-
-        String[] entries = targetFile.list();
-        if (entries != null && entries.length != 0) {
-            for (String s : entries) {
-                File currentFile = new File(targetFile.getPath(), s);
-                if (currentFile.isDirectory()) {
-                    delete(currentFile);
-                } else if (!currentFile.delete()) {
-                    return false;
-                }
-            }
+        try {
+            Files.walk(targetFile)
+                    .sorted(Comparator.reverseOrder()) // To ensure the root directory is the one deleted last
+                    .map(Path::toFile)
+                    .forEach(File::delete);
+        } catch (SecurityException e) {
+            String errMsg = "Failed to delete the file: '" + targetFile.toString() + "'. Permission denied.";
```

Comment:
Shall we use a string variable for - "Failed to delete the file: '" as it is redundant?

Reply:
Can do the same to other native implementations too (move, mkdirs).

Reply:
In Delete, it doesn't really make sense since this is just in 2 places